### PR TITLE
[RISCV] Remove vmv.s.x and vmv.x.s lmul pseudo variants

### PR DESCRIPTION
**Pull request by lukel97**
vmv.s.x and vmv.x.s ignore LMUL, so we can replace the PseudoVMV_S_X_MX and
PseudoVMV_X_S_MX with just one pseudo each. These pseudos use the VR register
class (just like the actual instruction), so we now only have TableGen patterns for vectors of LMUL <= 1.
We now rely on the existing combines that shrink LMUL down to 1 for vmv_s_x_vl (and vfmv_s_f_vl). We could look into removing these combines later and just inserting the nodes with the correct type in a later patch.

The test diff is due to the fact that a PseudoVMV_S_X/PsuedoVMV_X_S no longer
carries any information about LMUL, so if it's the only vector pseudo
instruction in a block then it now defaults to LMUL=1.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp:3367]: This is needed for the implicit_def -> noregister peephole: Without it we lose some CSE when we have (vmv_s_x IMPLICIT_DEF, ...) and (vmv_s_x (extract_subreg IMPLICIT_DEF, ...), ...)

File: llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td:7421]: `is_group`->`isGroup`

File: llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td:7426]: I don't have a name here but we don't use underlines for local variables usually.

**Review by wangpc-pp — COMMENTED** [review]
Thanks! I have never thought it would be so easy.

File: llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td
Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td:6764]: Sorry, not sure what happened here

**Comment by lukel97** [comment]
> Thanks! I have never thought it would be so easy.

A bit of context I should probably mention, it's easy to do now because https://github.com/llvm/llvm-project/pull/69788 has landed. Prior to that removing the LMUL information from the pseudos resulted in a lot of extra vsetvli toggles

File: llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td:7428]: How about some meaningful names like `vmv`, `pseudoVMV`?

**Review by wangpc-pp — APPROVED** [review]
LGTM with nits. And I need more eyes on the test diffs since they are large.

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-int-interleave.ll
Comment (lukel97) [inline llvm/test/CodeGen/RISCV/rvv/fixed-vectors-int-interleave.ll:492]: Here's a case where a backwards pass of #69259 would help.

**Comment by lukel97** [comment]
Ping for another set of eyes over the test diffs

File: llvm/test/CodeGen/RISCV/rvv/extractelt-int-rv32.ll
Comment (michaelmaitland) [inline llvm/test/CodeGen/RISCV/rvv/extractelt-int-rv32.ll:11]: Is there agreement that m1 compared to mf8 is not a regression? Some microarchitectures have same behavior for M1 and fractional LMUL. I don't know whether there is any microarchitectures that can optimize on the fractional LMUL cases.

**Comment by topperc** [comment]
Do we still compile this with a single vsetvli after this change https://godbolt.org/z/bvGd46zWa

File: llvm/test/CodeGen/RISCV/rvv/extractelt-int-rv32.ll
Comment (lukel97) [reply llvm/test/CodeGen/RISCV/rvv/extractelt-int-rv32.ll:11]: We already increase LMUL for vmv.s.x and vmv.x.s in RISCVInsertVSETVLI, albeit it to avoid a vsetvli. Perhaps that already sets a precedent.

Comment (topperc) [reply llvm/test/CodeGen/RISCV/rvv/extractelt-int-rv32.ll:11]: The vmv.x.s instruction doesn't really use the VL or the LMUL. It doesn't require the vector register to be naturally aligned to the LMUL. It only uses SEW to know how many bits to extract and where to sign extend in the scalar register. This makes it different than other instructions so it's hard to know for sure.

**Comment by lukel97** [comment]
> Do we still compile this with a single vsetvli after this change https://godbolt.org/z/bvGd46zWa

Just checked, there's no change

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-masked-gather.ll
Comment (topperc) [inline llvm/test/CodeGen/RISCV/rvv/fixed-vectors-masked-gather.ll:7408]: Where did m2 come from?

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-unaligned.ll
Comment (topperc) [inline llvm/test/CodeGen/RISCV/rvv/fixed-vectors-unaligned.ll:450]: Any fix for this?

**Comment by topperc** [comment]
What would it take to remove the INSERT_SUBREG/EXTRACT_SUBREG from tblgen? Can we avoid creating the nodes with non-LMUL types in the first place?

**Comment by lukel97** [comment]
> What would it take to remove the INSERT_SUBREG/EXTRACT_SUBREG from tblgen? Can we avoid creating the nodes with non-LMUL types in the first place?

I think we might be able to wrap all the places where we emit VMV_S_X_VL/VMV_X_S_VL in RISCVISelLowering with INSERT_SUBVECTOR/EXTRACT_SUBVECTORs instead. I'll give that a try

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-masked-gather.ll
Comment (lukel97) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-masked-gather.ll:7408]: My guess is that it tried to touch up the SEW/LMUL ratio so it's the same as bb.0. But I'm not sure why it wasn't able to emit the x0,x0 form here, it looks like it should be ok to use it.

File: llvm/lib/Target/RISCV/RISCVInstrInfoVVLPatterns.td
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVInstrInfoVVLPatterns.td:2762]: Move this comment down to above new foreach.

File: llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td:7418]: I think you can leave these as patterns and just drop the MX suffix?  Would be more obviously NFC.

File: llvm/test/CodeGen/RISCV/rvv/extractelt-int-rv32.ll
Comment (preames) [reply llvm/test/CodeGen/RISCV/rvv/extractelt-int-rv32.ll:11]: I don't know of any reason to think increasing to m1 here will have any effect on known processors, but there's a lot of uncertainty here.  I will note that defaulting to smallest-legal-fractional-for-sew in insert vsetvli insertion on this instruction probably wouldn't be too painful, so we can likely reverse this easily if needed.  (Luke, might be worth playing with that just to reduce the diff.)

**Review by preames — COMMENTED** [review]
One concerning possible regression, otherwise looks straight forward.

File: llvm/test/CodeGen/RISCV/rvv/extractelt-int-rv32.ll
Comment (michaelmaitland) [reply llvm/test/CodeGen/RISCV/rvv/extractelt-int-rv32.ll:11]: If there was another vector instruction in the sequence that needed some different LMUL other than m1, then that LMUL would have been used instead of m1, since vmv.s.x and vmv.x.s has no impact on LMUL anymore. So I am not worried about the possibility of an extra vsetvli instruction being inserted.

I also don't know of any hardware that has differing performance on fractional LMUL compared to m1. For that reason, I think we could probably hold off on defaulting to the smallest-legal-fractional-for-sew in vsetvli insertion until there was a use case that made it worthwhile.

**Comment by topperc** [comment]
> > Do we still compile this with a single vsetvli after this change https://godbolt.org/z/bvGd46zWa
> 
> Just checked, there's no change

This was what I should have sent, the integer version https://godbolt.org/z/f9bzqz6aq but it already has a vsetvli toggle. :(

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-masked-scatter.ll
Comment (wangpc-pp) [inline llvm/test/CodeGen/RISCV/rvv/fixed-vectors-masked-scatter.ll:1924]: Is this a regression here?

Comment (lukel97) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-masked-scatter.ll:1924]: Yeah, I meant to flag this earlier sorry. These were introduced with the last commit to avoid extract_subreg/insert_subreg d390274cf8295c5e40ae06d7308ceace0c1ee82b

It's because this combine no longer kicks in now that the vmv_x_s has its operand wrapped in an extract_subvector:
https://github.com/llvm/llvm-project/blob/51c351f49ed1e1406543fcadc7f6d3a4272e74bd/llvm/lib/Target/RISCV/RISCVISelLowering.cpp#L15172-L15190

**Comment by lukel97** [comment]
> > > Do we still compile this with a single vsetvli after this change https://godbolt.org/z/bvGd46zWa
> > 
> > 
> > Just checked, there's no change
> 
> This was what I should have sent, the integer version https://godbolt.org/z/f9bzqz6aq but it already has a vsetvli toggle. :(

I think this is because we don't do the sew/lmul ratio trick in the backwards local postpass. I ran #71657 on this and it's able to remove the toggle

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-mask-buildvec.ll
Comment (lukel97) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-mask-buildvec.ll:555]: This is new to me, but this might actually be an intentional optimisation: In `RISCVInstrInfo::copyPhysRegVector` we actually try to convert vmxr.vs to their vmv.v.v counterparts https://reviews.llvm.org/D103510
Presumably this triggers the change because the LMUL now exactly lines up.
Looks like there's a flag to control this too with `-riscv-prefer-whole-register-move`. Maybe this should become a subtarget feature?

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-masked-scatter.ll
Comment (lukel97) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-masked-scatter.ll:1924]: I could add a check in here like 
```c++
      SDValue Src = Val.getOperand(0);
      if (Src.getOpcode() == ISD::EXTRACT_SUBVECTOR && isNullConstant(Src.getOperand(1)))
	Src = Src.getOperand(0);
```
but then we actually end up increasing LMUL in some cases, e.g.

```diff
 define void @store_vfmv_f_s_nxv8f64(<vscale x 8 x double>* %x, double* %p) {
 ; CHECK-LABEL: store_vfmv_f_s_nxv8f64:
 ; CHECK:       # %bb.0:
 ; CHECK-NEXT:    vl8re64.v v8, (a0)
-; CHECK-NEXT:    vsetivli zero, 1, e64, m1, ta, ma
+; CHECK-NEXT:    vsetivli zero, 1, e64, m8, ta, ma
 ; CHECK-NEXT:    vse64.v v8, (a1)
 ; CHECK-NEXT:    ret
   %a = load <vscale x 8 x double>, <vscale x 8 x double>* %x
   %b = call double @llvm.riscv.vfmv.f.s.nxv8f64(<vscale x 8 x double> %a)
   store double %b, double* %p
   ret void
 }
```

This extra vsetvli actually decreases LMUL: Does this have any effect on performance on a vse32e if VL is the same?
cc @preames @topperc

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-unaligned.ll
Comment (lukel97) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-unaligned.ll:450]: #72352 should fix this

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-masked-scatter.ll
Comment (preames) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-masked-scatter.ll:1924]: I don't think this regression is particularly concerning.  Given there can be costs associated with a higher LMUL on an operation, it's not even clear to me that having the extra vsetvli here is actually a regression in practice.  I don't think this is worth trying to change at this time.

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-mask-buildvec.ll
Comment (topperc) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-mask-buildvec.ll:555]: The transform probably has more merit for higher LMUL if hardware optimizes number of uops/cycles based on VL and VL is smaller than the LMUL.

That transform would be bad for CPUs that implement move elimination for whole register moves, but not for vmv.v.v.

File: llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td:7414]: Why do we still need this pattern?

**Comment by topperc** [comment]
Sorry Luke, I caused a conflict with c281a6add55a861a16eabe3bb8b436f63ef8e0cb because I didn't realize this patch had picked up FP.

File: llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td
Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td:7414]: We don't, I originally removed it but added it back in as per @preames suggestion https://github.com/llvm/llvm-project/pull/71501#discussion_r1392997610 (looks like GitHub isn't showing the comment inline anymore though 😬)

File: llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td
Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td]: What's the advantage of having a dead pattern? Or are you suggesting to not remap int_riscv_vfmv_s_f to RISCVISD::VFMV_S_F_VL during lowering?

File: llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td:244]: Doesn't this need to be moved to VVLPatterns.td?

Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVInstrInfoVPseudos.td:244]: Whoops sorry, didn't notice it came from there

File: llvm/test/CodeGen/RISCV/rvv/extractelt-int-rv32.ll
Comment (lukel97) [reply llvm/test/CodeGen/RISCV/rvv/extractelt-int-rv32.ll:11]: I tried out defaulting to mf8 for vmv.s.x and vmv.x.s here https://github.com/lukel97/llvm-project/commit/59bd2221c1970f5c1254d26196c6990978b88751, and then rebasing this branch on top of it here https://github.com/lukel97/llvm-project/commit/1097e21545607d3aaaa00f821127d666f3da9ac2

Doing it this way moves most of the test diffs into that prior commit, but we still see the same missing vsetvli transforms etc. in it

**Comment by topperc** [comment]
What's the status of this patch?

**Comment by lukel97** [comment]
> What's the status of this patch?

Depending if we want some of the regressions addressed where we lose the x0,x0 vsetvlis, waiting on #74040

**Comment by lukel97** [comment]
Rebased now that #74040 is landed, test diffs should be smaller now. 

The one remaining regression is where we lose the x0, x0 form when a vmv.s.x/vmv.x.s is the first instruction in the basic block: https://github.com/llvm/llvm-project/pull/71501/files#diff-10cadf020a900ddef6073743714540c0dd895bb1eabc9555c9fe6a0004f9d15bL358-L361

I think the correct fix for this is to perform SEW/LMUL ratio adjustment in the backwards pass (Or generally improve how we emit the x0, x0 form). But this doesn't seem to be a common occurrence, and I would prefer to leave it to a follow up patch.

**Comment by preames** [comment]
I happened to talk with Craig offline yesterday and we discussed this patch a bit.  Craig had previous raised a question above which appears to have gotten lost in the review discussion.  He posted https://github.com/llvm/llvm-project/pull/75544 to land the test cases in tree, and this patch needs to be rebased over that one, and the fix for the issue flagged in my comment as I believe that should resolve some diffs in this one.

I'd also like to request that you split this patch into the integer subset and the floating point subset.  The original patch was integer only, and some of the confusion on the review is caused by the addition of the floating point without it being clearly called out.

**Comment by lukel97** [comment]
Rebased after #76699

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**
